Fix potential race condition causing BadVersion

cmelchior opened the pull request:
Fixes #2417

cmelchior commented:
I believe this PR should fix the race condition, but we are now getting a segmentation fault in core in `TableQuery.nativeImportHandoverTableViewIntoSharedGroup` which baffles me a bit :unamused:

finnschiermer commented on realm/realm-jni/src/io_realm_internal_TableQuery.cpp:
```diff
@@ -1037,24 +1037,26 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFind(
         LangBindHelper::advance_read(*SG(bgSharedGroupPtr));
+    } else {
+        try {
+            LangBindHelper::advance_read(*SG(bgSharedGroupPtr), handoverQuery->version);
```
At this point in time, it may not be safe to access the handoverQuery. It is moved() at line 1045, so the access to version here is not well-defined.

cmelchior replied:
Ah yes, good point

cmelchior replied:
Unfortunately that didn't seem to the root cause of the crash

nhachicha commented:
updated the PR please review @cmelchior @tgoyne @finnschiermer

cmelchior commented on realm/realm-jni/src/io_realm_internal_TableQuery.cpp:
```diff
@@ -1037,23 +1037,23 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFind(
-    if (isDifferentVersions) {
+    } else if (SG(bgSharedGroupPtr)->get_version_of_current_transaction() != handoverQuery->version) {
+        SG(bgSharedGroupPtr)->end_read();
         SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
```
This line can still cause a BadVersion which is currently not getting caught and will crash. 
Also, right now our executor service has more than 1 thread, by restricting it to 1 thread It seems we can control the concurrency enough for a stable unit tests.

I have an idea for fixing it + actually making a proper unit test for this. Update incoming

cmelchior commented:
@nhachicha I updated the PR. We didn't properly handle BadVersionException in our findAllAsync methods from what I could tell. In a lot of cases we treated that exception as fatal (even though a retry was imminent). I think I fixed it, but still trying to wrap my head around testing it in a sane way. So far I have done manual testing but stubbing out values.

Please review, especially around retrying and batch updates.

cmelchior commented:
Closing this PR as it should be merged into `releases` instead.

jcj52436999 commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  jcjacksonengineering@gmail.com
    Domain joe-jackson.net has exceeded the max emails per hour (101/100 (101%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext1.iad.github.net ([192.30.252.192]:53611 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by rs5.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.86_1)
	(envelope-from <noreply@github.com>)
	id 1afmIx-001w0H-88
	for jcjackson@joe-jackson.net; Tue, 15 Mar 2016 06:34:23 -0400
Date: Sun, 13 Mar 2016 15:14:53 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1457907293;
	bh=B4E5IMH1d8hKnaUh3tuakjox6lbrGaZ+sz+h02bMWBk=;
	h=From:Reply-To:To:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=W3CQVkLf3kd75gt3AdXGH4OIPG3pGsAl4DUY+GuaHRLVY+BKmu4Wm2GMFbxtm6Qcw
	 Sfvb22kd8UnA0EDjjWZ1k3INb9qUc/u5XlSpEhYz6uHKFDzaByOeyCbKcnu+hROxPo
	 5br02SI41HwdCeuPCdh+XZwE2vwC2zjo8jyy53K8=
From: Nabil Hachicha <notifications@github.com>
Reply-To: realm/realm-java <reply@reply.github.com>
To: realm/realm-java <realm-java@noreply.github.com>
Message-ID: <realm/realm-java/pull/2426/c196063929@github.com>
In-Reply-To: <realm/realm-java/pull/2426@github.com>
References: <realm/realm-java/pull/2426@github.com>
Subject: Re: [realm-java] Fix potential race condition causing BadVersion
 (#2426)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_56e5e65d92945_63003ffd6ff4b29c11164d";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: nhachicha
X-GitHub-Recipient: jcj52436999
List-ID: realm/realm-java <realm-java.realm.github.com>
List-Archive: https://github.com/realm/realm-java
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+000f94f79dd9d5909860983fbd348888f628a856187611e692cf0000000112fda85d92a169ce085a9f72@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AA-U9_yYd44L7Bfa7THs7sCF4lh5bxPjks5ptIxdgaJpZM4Hulvl>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: jcjackson@joe-jackson.net


----==_mimepart_56e5e65d92945_63003ffd6ff4b29c11164d
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

updated the PR please review @cmelchior @tgoyne @finnschiermer 

---
Reply to this email directly or view it on GitHub:
https://github.com/realm/realm-java/pull/2426#issuecomment-196063929
----==_mimepart_56e5e65d92945_63003ffd6ff4b29c11164d
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p>updated the PR please review <a href="https://github.com/cmelchior" class="user-mention">@cmelchior</a> <a href="https://github.com/tgoyne" class="user-mention">@tgoyne</a> <a href="https://github.com/finnschiermer" class="user-mention">@finnschiermer</a> </p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br>Reply to this email directly or <a href="https://github.com/realm/realm-java/pull/2426#issuecomment-196063929">view it on GitHub</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AA-U90WKDLu-k9h8yrMyFIvgUQ3b0OiAks5ptIxdgaJpZM4Hulvl.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
<div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
  <link itemprop="url" href="https://github.com/realm/realm-java/pull/2426#issuecomment-196063929"></link>
  <meta itemprop="name" content="View Pull Request"></meta>
</div>
<meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

----==_mimepart_56e5e65d92945_63003ffd6ff4b29c11164d--
